### ISSUE TITLE
reentering password for google account on android devices

### ISSUE THREAD

**Issue by m0gg** [diraimondo/gmusicproxy#34]
I'm using your gmusicproxy in an application to mirror my playlists and used mp3 which works really nice.
Anyway i occasionally need to reenter my password on every android device related to this account. The message states that i have to do this due to a change in account settings.
I just wondered if you have encountered this too?

**Comment by diraimondo** [comment]
It happened to me in the past but just few times. I don't know if it is related to gmusicproxy. Are you using the two-phases authentication in Google services (so, with an application-specific password)? Are you using the id of a real android/ios device?

**Comment by m0gg** [comment]
No i don't use two-phase auth. But i guess it's related to gmusicproxy, if i disable the cronjob it won't happen.

**Comment by diraimondo** [comment]
So it is systematic for you (it happens every time gmusicproxy is used)... This happened so sporadically for me in the past that I was not able to relate it to gmusicproxy.
What version of gmusicproxy and gmusicapi are you using?

**Comment by m0gg** [comment]
https://gist.github.com/m0gg/952f93ece47bc9457e30
Is the full pip list

gmusicapi (6.0.0)
gmusicproxy (1.0.2)

**Comment by diraimondo** [comment]
I have no idea where could be the problem. No other people reported such kind of problems. Maybe @simon-weber can help us.

**Comment by simon-weber** [comment]
Huh, I don't know either; it's the first I've heard of this. You could try upgrading to the newest gmusicapi, but I'm not aware of any fixes in this area.

**Comment by diraimondo** [comment]
This side-effect looks not clearly related to gmusicproxy/gmusicapi so I'm going to close this report. Feel free to reopen it if you get more evidences or a solution.

**Comment by Tadly** [comment]
I have/had the same issue. I'm still not exactly sure why this is **but** I know it happens when you send to many login-requests.
After about 30 logins, I get logged out on all my devices.

The solution for now is to save the *authtoken* and instead of doing another login, set the authtoken again. (check the script below on how exactly I did that)

To test this I created a small (and ugly) script. If someone feels like testing it, set ***testcase*** to either **1** or **2** and ***android_id*** to one of your ids (or use Mobileclient.FROM_MAC_ADDRESS)

If ***testcase*** is set to ***1*** I get logged out, if set to ***2*** I don't (and using the api still works)

```python
from getpass import getpass
from gmusicapi import Mobileclient

testcase   = 2
runcount   = 30
android_id = ''
username   = raw_input('Username: ')
password   = getpass()


# Spam "real" login/logout requests
if testcase == 1:
    for i in range(runcount):
        # Recreating the client-object on every request would be ruffly the
        # same behaviour as with a kodi-addon. That's why I do it here as well
        api = Mobileclient()

        print('Login/Logout test: %s' % i)
        api.login(username, password, android_id)
        print('    Testrequest response size: %s' % len(api.get_registered_devices()))
        api.logout()

# Login only once the get a "authtoken" and reuse it for fututre requests
if testcase == 2:
    api = Mobileclient()
    api.login(username, password, android_id)
    authtoken = api.session._authtoken
    api.logout()
    for i in range(runcount):
        # Recreating the client-object on every request would be ruffly the
        # same behaviour as with a kodi-addon. That's why I do it here as well
        api = Mobileclient()

        print('Login/Logout test: %s' % i)
        api.android_id = android_id
        api.session._authtoken = authtoken
        api.session.is_authenticated = True
        print('    Testrequest response size: %s' % len(api.get_registered_devices()))
        api.logout()
```

**Comment by diraimondo** [comment]
I've tested your code with my daily-used application-specific credentials. With testcase 1 I can see the successfully logins until about 26 attempts in row. I guess there is a limit per time. After some minutes I can login again with the same credentials. 
After these testings I can continue to use my Google-related sessions on my browser. I also tried to close-reopen it. Sorry, but this testcase doesn't create the problem in my environment.

The testcase 2 is interesting but:
- as long as this is not a problem, this is not necessary;
- with can't play with the internal variables of the objects of gmusicapi; @simon-weber  should implement this kind of authentication caching first.

**Comment by Tadly** [comment]
@diraimondo with testcase 1 the session on my browser work too but on all android-devices they don't.  Either I automatically get a notification saying I have to re-login or (if there is no notification just yet) as soon as I open a google-acc related app (e.g. Play Store etc.) I get prompted to re-enter my credentials.

Haven't tested it with *Mobileclient.FROM_MAC_ADDRESS* though.
Maybe using a real android-id is the problem here? (I prefer using a real id over the mac-address to make sure I don't add devices to my device-list on playmusic as you are limited in the amount of active devices and the amount of devices you can remove per year)

**Comment by m0gg** [comment]
Confirming it won't logout browser sessions. Happened only to my androids aswell.

**Comment by diraimondo** [comment]
Ok, I rechecked and this high-rate login attempts lock android devices associated to the account. I have to say that this doesn't look a bug of gmusicapi login system. This testcase can't be considered a regular usage pattern.
If @simon-weber thinks that it is worthy to implement the caching of authentication token in gmusicapi, I could support it too in gmusicproxy.

**Comment by Tadly** [comment]
@diraimondo true. I came across this issue while developing a addon for kodi. With kodi addons your script gets completely unloaded after performing one action which means I have to do a login after each user interaction.
While doing that my devices got locked quite often.

What I'm trying to say is that even with "normal" use this could/can happen.

But it should be clear now that this isn't a **bug** in that sense but more something that could be improved.

At least we know what the problem is now :)

@simon-weber would love to see authtoken caching in gmusicapi. For now I'll just do it myself ;)

**Comment by diraimondo** [comment]
@Tadly , because you opened a report on gmusicproxy project, I can guess that you are trying to use gmusicproxy as in-the-middle service, right?! Why gmusicproxy can't just run as an independent service and let kodi plugin to connect when it is required?

**Comment by Tadly** [comment]
@diraimondo actually I don't ^^
I've been asking simon about that issue (in irc) and he linked me to this issue saying that's all he knows about.
As no one knew what the issue was I though I test some stuff and provide as much info as possible to hopefully nail it down before opening another issue (on gmusicapi) that's basically the same ^^

**Comment by diraimondo** [comment]
@Tadly Sorry, I didn't note this issue was opened by another person. Sorry.

**Comment by Tadly** [comment]
@diraimondo No problem :)

**Comment by simon-weber** [comment]
> If @simon-weber thinks that it is worthy to implement the caching of authentication token in gmusicapi, I could support it too in gmusicproxy.

I don't think I'll support it transparently, but I can certainly document it and support logging in via an auth token -- ie, properly support https://github.com/diraimondo/gmusicproxy/issues/34#issuecomment-147359198 as a workaround.

**Comment by m0gg** [comment]
Thanks for your support. I encoutered this during development too. Unfortunately the logout kept happening while production too. That made me curious about it. But the pattern matches. My cronjob quits the proxy when done and fires up a new instance when run again.
